crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1190,128 +1200,28 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.extend(result);
 
         let value_ty = self.expression_ty(value);
+        let name_ast_id = name.scoped_ast_id(self.db, self.scope());
 
         let target_ty = match kind {
-            AssignmentKind::Sequence => self.infer_sequence_unpacking(target, value_ty, name),
+            AssignmentKind::Sequence => {
+                let target = definition.kind(self.db).as_unpack_target().unwrap();
+                let unpack = Unpack::new(
+                    self.db,
+                    self.file,
+                    definition.file_scope(self.db),
+                    target,
+                    value_ty,
+                    countme::Count::default(),
+                );
+                let unpacked = infer_unpack_types(self.db, unpack);
+                self.diagnostics.extend(unpacked.diagnostics());
+                unpacked.get(name_ast_id).unwrap_or(Type::Unknown)
+            }
```

Comment:
I'm not exactly a big fan of this; what I want here is to get the target expression for the unpacking and it should be a `AstNodeRef`. It doesn't really matter if we know where the target is coming from which we currently do using the `UnpackTarget` enum.

Another way would be to accept the `target: AstNodeRef<Expr>` parameter for `infer_assignment_definition` but then we'd need to update the `AssignmentDefinitionKind` to contain this target expression as `AstNodeRef` instead of just storing the target index:

https://github.com/astral-sh/ruff/blob/c03f9873a1e884a03e1ccfd5fa4848df5fae863f/crates/red_knot_python_semantic/src/semantic_index/definition.rs#L518-L524

And, then for the `ForStmtDefinitionKind`, we'd update the existing `target` field type to be `AstNodeRef<Expr>`:

https://github.com/astral-sh/ruff/blob/c03f9873a1e884a03e1ccfd5fa4848df5fae863f/crates/red_knot_python_semantic/src/semantic_index/definition.rs#L563-L568

And, similarly for other nodes where unpacking is supported.